### PR TITLE
ARROW-13327: [C++][Python] Improve consistency of explicit C++ types in PyArrow files

### PR DESCRIPTION
**Pull request by edponce**
This PR updates C++ and PyArrow compute bindings for FunctionOptions and makes their API/implementations more consistent.
Also, positional and keyword-only parameters in PyArrow FunctionOptions are updated.

### CODE REVIEW

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:728]: Is there missing a return value here?

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:849]: Is there a reason to remove them both here and in the init just below?

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:874]: This is changing the name of the keyword in the public function?

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:903]: I would keep this (it's to force `boundscheck` to be a keyword argument, and never positional argument)

(and the same for some cases below)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks, added some comments.

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:849]: Yes, they were removed deliberately. The option values are passed to the constructors of the corresponding FunctionOptions which are defined in [includes/libarrow.pxd](https://github.com/apache/arrow/blob/master/python/pyarrow/includes/libarrow.pxd) and expose the explicit C types. Invalid types from public functions in .pyx get "validated" during initialization in constructors. If the types are included in the `cdef _XXXOptions`, then inputs would be "validated" here, basically one line of code earlier. Duplicating types in the `_FunctionOptions` methods in `_compute.pyx` does not adds any different functionality, but requires developers/maintainers to ensure consistency between these files and the C++ option definitions. I think removing them from .pyx file is a better approach.

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:874]: Yes, for symmetry it would have been better to not include the `_behavior` part but given that this was existing code and there might already be users expecting this API, I will revert this change to how it was before this PR.

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:903]: I remove all the keyword arguments simply to get a code review like yours. How do we decide what options should be positional and which ones should be keyword-only?

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [reply python/pyarrow/_compute.pyx:826]: OK, I see, thanks for the explanation.

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:881]: Ok, I started doing that (still WIP), but also noticed that in C++ FunctionOptions, there are inconsistencies between the number of constructors, the availability of a `Defaults()` method, and data members. I am looking into this (need to reach out to other developers) to check what changes we are allowed to make at this time.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:234]: Call this `_kind_dict`? This isn't a public API.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:288]:
```python
try:
    return self.kind_dict[c_kind]
except KeyError:
    raise NotImplementedError("Unknown Function::Kind")
```

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:724]: This change doesn't look desirable to me. The Python build-in [round](https://docs.python.org/3/library/functions.html#round) function takes  positional arguments.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:736]: Similarly, this change doesn't look desirable.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:778]: Similarly, this doesn't seem desirable. I don't think there's any ambiguity as to what the second argument is.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:831]: Same here. Why would `stop` be a keyword-only parameter?

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:836]: `_null_selection_dict`

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:851]: Having to type `null_selection_behavior` is rather annoying...

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:932]: Making this keyword-only doesn't seem useful, IMHO.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:961]: `n` could very well be positional.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:1130]: Similarly, this doesn't seem to increase usability or readability.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:1146]: Well, it should be clear that the first argument to `select_k` is the `k` number, no?

**Review by pitrou — CHANGES_REQUESTED** [review]
Some of these changes look a bit unwarranted to me. When there is little ambiguity as to what a parameter means, there is no reason to make it keyword-only.

@jorisvandenbossche Any thoughts on these changes?

**Comment by edponce** [comment]
The FunctionOptions that were modified w.r.t. to positional/keyword-only args are:
```python
MatchSubstringOptions(self, pattern, *, ignore_case)
ReplaceSubstringOptions(self, pattern, replacement, *, max_replacements)
ScalarAggregateOptions(self, *, skip_nulls, min_count)
ModeOptions(self, n, *, skip_nulls, min_count)
SetLookupOptions(self, value_set, *, skip_nulls)
DayOfWeekOptions(self, *, one_based_numbering, week_start)
NullOptions(self, *, nan_is_null)
SplitPatternOptions(self, pattern, *, max_splits, reverse)
ArraySortOptions(self, order)
```

File: python/pyarrow/includes/libarrow.pxd
Comment (pitrou) [inline python/pyarrow/includes/libarrow.pxd:2087]: It seems these `cdef` changes may have broken AppVeyor. Can you take a look?

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:546]: This is a private function, prefix it with an underscore?

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:685]: Should this one become keyword-only? `ElementWiseAggregateOptions(true)` isn't very explicit.

File: python/pyarrow/tests/test_compute.py
Comment (pitrou) [inline python/pyarrow/tests/test_compute.py:1942]: Nit, but the quotes around "sort order" look a bit gratuitous?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the followup. I posted a couple additional comments. Most notably though, AppVeyor CI should be fixed :-)

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:546]: Do you think we should include it as a private method of `FunctionOptions` or leave it as a free standing function?

Comment (pitrou) [reply python/pyarrow/_compute.pyx:546]: Can leave it as a free standing function.

File: python/pyarrow/_compute.pyx
Comment (edponce) [inline python/pyarrow/_compute.pyx:694]: `RoundMode` and `SortOrder` are scoped enums but they are not fully supported (in terms of concise usage) until Cython 3. Arrow currently builds for Cython >= 0.29.x, so I left *TODO* comments ([here](https://github.com/apache/arrow/pull/11147/files#diff-439f91c435cc8136d40eaba8c168aaa1cec00f08b10ba92ce376e47f29f81814R690) and [here](https://github.com/apache/arrow/pull/11147/files#diff-439f91c435cc8136d40eaba8c168aaa1cec00f08b10ba92ce376e47f29f81814R1114)) to describe the intent here for working with scoped enums. Also, left *TODO* comments ([here](https://github.com/apache/arrow/pull/11147/files#diff-48e87fef366d65330100bf1703a542b8cb447575b6f8c3a3617741ce57854d4aR2090) and [here](https://github.com/apache/arrow/pull/11147/files#diff-48e87fef366d65330100bf1703a542b8cb447575b6f8c3a3617741ce57854d4aR1829)) for defining scoped enums in `includes/libarrow.pxd`.

This is the first time I embed *TODO* comments in Arrow but think they are worthwhile. I can also keep track of these notes locally or should this go into a future JIRA issue? @pitrou please advise.

Comment (pitrou) [reply python/pyarrow/_compute.pyx:694]: You can open a JIRA for this. Of course it's a pretty minor issue.

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:546]: This will give e.g.:
```"foo" is not a valid count mode keyword```
IMHO it would read better as:
```"foo" is not a valid count mode value```

File: python/pyarrow/_compute.pyx
Comment (pitrou) [inline python/pyarrow/_compute.pyx:992]: If the type is the problem, then probably `TypeError` should be raised, not `ValueError`.

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:913]:
```suggestion
    def __init__(self, field_names, *, field_nullability=None,
```

?

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [reply python/pyarrow/_compute.pyx:546]: Or could also leave out the final word entirely: `"foo" is not a valid count mode`

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:1025]: (another styling nit: note that some formatters (eg black) will actually prefer how it was before)

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:1182]:
```suggestion
    def __init__(self, q=0.5, *, interpolation="linear", skip_nulls=True,
```

? 

(since `q` is not a very descriptive keyword name anyway, it's not necessarily useful to have to spell it out)

File: python/pyarrow/compute.py
Comment (jorisvandenbossche) [inline python/pyarrow/compute.py:679]:
```suggestion
    sort_keys : list
```

Or "list-like" in you prefer. Since it supports any sequence, I don't think it's needed to specifically mention "tuple" in addition to list.

(reason this catched my attention is actually because I first misread it as "list of tuple", while `sort_keys` in general follow that structure, while here it should be just be a list of strings)

File: python/pyarrow/compute.py
Comment (jorisvandenbossche) [inline python/pyarrow/compute.py:725]:
```suggestion
    sort_keys : list
```

**Comment by jorisvandenbossche** [comment]
> When there is little ambiguity as to what a parameter means, there is no reason to make it keyword-only.
>
> @jorisvandenbossche Any thoughts on these changes?

Yeah, we had a bit of discussion about it on Zulip, where the it was suggested to follow "required args -> positional, optional args -> keyword only", but so that clearly isn't a good rule, as I agree with your comments (so my initial "make a judgement call case by case based on whether you could understand it without the current" is better).

> The FunctionOptions that were modified w.r.t. to positional/keyword-only args are:

I think the changes listed at https://github.com/apache/arrow/pull/11147#issuecomment-922733817 all look good now.

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:335]: I agree, I personally prefer whitespace when new context follows, but I have noticed that a lot of Arrow code is clumped together. I reverted this change.

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:913]: `MakeStructOptions` was updated recently with additional mandatory parameters. All arguments are mandatory (see [C++ definition](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/api_scalar.h#L300-L303)), so I think we should leave them positional here as well.

Comment (edponce) [reply python/pyarrow/_compute.pyx]: Oops, I now noticed that `MakeStructOptions` does have constructor with only the first parameter (`field_names`). I applied the suggested change (keyword-only for optional parameters).

File: python/pyarrow/_compute.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:1078]:
```suggestion
            _raise_invalid_function_option(ambiguous, "'ambiguous' value")
        if nonexistent not in self._nonexistent_map:
            _raise_invalid_function_option(nonexistent, "'nonexistent' value")
```

In this case there needs to be an extra word I think (otherwise you get " ... is not a valid ambiguous" as error message)

**Review by jorisvandenbossche — APPROVED** [review]

File: python/pyarrow/_compute.pyx
Comment (edponce) [reply python/pyarrow/_compute.pyx:1078]: What about "'ambiguous' timezone assumption"?

Comment (jorisvandenbossche) [reply python/pyarrow/_compute.pyx:1078]: The "ambiguous" is not the about the timezone, but the timestamp. So something like `'ambiguous' timestamp handling` would be OK as well I think

Comment (edponce) [reply python/pyarrow/_compute.pyx]: Thanks! This change has been added.

**Review by pitrou — APPROVED** [review]
+1, thank you @edponce

**Pull request closed without merging**
